ui/src/funding/pool.ts

```diff
@@ -0,0 +1,335 @@
+import * as svelteStore from "svelte/store";
+
+import * as contract from "./contract";
+import * as transaction from "../transaction";
+import * as validation from "../validation";
+
+import { Wallet, Account, Status as WalletStatus } from "../wallet";
+import * as remote from "../remote";
+import { BigNumber, BigNumberish, ContractTransaction, ethers } from "ethers";
+
+export const store = svelteStore.writable<Pool | null>(null);
+
+export interface Pool {
+  data: remote.Store<PoolData>;
+
+  // Get the account that owns this pool. Should be the connected wallet account.
+  getAccount: () => Account | undefined;
+
+  // Onboard the user's pool with the intial values
+  onboard(
+    topUp: BigNumberish,
+    amountPerBlock: BigNumberish,
+    receivers: Receivers
+  ): Promise<void>;
+
+  // Update the contribution per block and the list of receivers.
+  updateSettings(amountPerBlock: string, receivers: Receivers): Promise<void>;
+
+  // Adds funds to the pool. Returns once the transaction has been
+  // included in the chain.
+  topUp(value: BigNumberish): Promise<void>;
+
+  // Withdraw outgoing balance funds to the connected wallet.
+  // Returns once the transaction has been included in the chain.
+  withdraw(value: BigNumberish): Promise<void>;
+
+  // Withdraw all the outgoing balance funds to the connected wallet.
+  withdrawAll(): Promise<void>;
+
+  // Collect funds the user has received up to now from givers and
+  // transfer them to the users account.
+  collect(): Promise<void>;
+
+  // Approve the ERC-20 token allowance, which permits
+  // the pool to access said type of token from the user.
+  approveErc20(): Promise<void>;
+}
+
+// All the data representing a pool.
+export interface PoolData {
+  // The remaining balance of this pool.
+  balance: BigNumberish;
+  // The total amount to be disbursed to all receivers with each block.
+  amountPerBlock: BigNumberish;
+  // The list of addresses that receive funds from the pool.
+  receivers: Receivers;
+  // Funds that the user can collect from their givers.
+  collectableFunds: number;
+  // The ERC-20 token allowance. 0 means that the allowance was not
+  // granted or that it was fully spent.
+  erc20Allowance: BigNumberish;
+}
+
+/* Receivers */
+export type Receivers = Map<Address, ReceiverStatus>;
+
+export type Address = string;
+export type Weight = BigNumberish;
+
+export enum ReceiverStatus {
+  // The receiver is being added
+  Added = "Added",
+  // The receiver is being removed
+  Removed = "Removed",
+  // The receiver is present already
+  Present = "Present",
+}
+
+export function make(wallet: Wallet): Pool {
```

Comment:
:+1: on using an explicit `wallet` dependency rather than importing it from the wallet module.